### PR TITLE
Add support for "Auto" for control size

### LINKED ISSUE #4403: Can't use "Auto" for FrameworkElement Width & Height
## Current behavior

When a `Style` setter targets `Width` or `Height` property and sets it to `Auto`, application crashes at runtime due to conversion issue.

## Expected behavior

Should not crash, `Auto` is equivalent to `NaN` for `Width` and `Height`

## How to reproduce it (as minimally and precisely as possible)

    <Setter Property="Width" Value="Auto" />

Maybe the problem occurs when directly in XAML (did not verify yet):

    <Grid Width="Auto" />

## Workaround

Set `NaN` directly.

## Environment

Nuget Package:

- [x] Uno.UI / Uno.UI.WebAssembly / Uno.UI.Skia
- [ ] Uno.WinUI / Uno.WinUI.WebAssembly / Uno.WinUI.Skia
- [ ] Uno.SourceGenerationTasks
- [ ] Uno.UI.RemoteControl / Uno.WinUI.RemoteControl
- [ ] Other: 

Nuget Package Version(s):

Affected platform(s):

- [x] iOS
- [x] Android
- [x] WebAssembly
- [x] WebAssembly renderers for Xamarin.Forms
- [x] macOS
- [x] Skia
  - [x] WPF
  - [x] GTK (Linux)
  - [x] Tizen
- [ ] Windows
- [ ] Build tasks
- [ ] Solution Templates

### PR DESCRIPTION
**Pull request by carldebilly**
Fixes #4403

# Feature
It is now possible to set bound values (and style setters) as 'Auto' in dependency properties.

## PR Checklist

Please check if your PR fulfills the following requirements:

- [ ] Docs have been added/updated which fit [documentation template](https://github.com/unoplatform/uno/blob/master/doc/.feature-template.md) (for bug fixes / features)
- [X] [Unit Tests and/or UI Tests](https://github.com/unoplatform/uno/blob/master/doc/articles/uno-development/working-with-the-samples-apps.md) for the changes have been added (for bug fixes / features) (if applicable)
- [ ] Validated PR `Screenshots Compare Test Run` results.
- [ ] Contains **NO** breaking changes
- [X] Associated with an issue (GitHub or internal) and uses the [automatic close keywords](https://help.github.com/en/github/managing-your-work-on-github/linking-a-pull-request-to-an-issue).
- [X] Commits must be following the [Conventional Commits](https://www.conventionalcommits.org/en/v1.0.0/#summary) specification.

### CODE REVIEW

File: src/Uno.UI/DataBinding/BindingPropertyHelper.FastConvert.cs [hunk 1]
```diff
@@ -334,6 +339,240 @@ private static bool FastStringConvert(Type outputType, string input, ref object
 			return false;
 		}
 
+		private static bool FastStringToNumericConvert(Type outputType, string input, ref object output)
+		{
+			const NumberStyles numberStyles = NumberStyles.AllowDecimalPoint | NumberStyles.AllowLeadingSign | NumberStyles.Float | NumberStyles.AllowLeadingWhite | NumberStyles.AllowTrailingWhite;
+
+			if (outputType == typeof(double))
```
Comment (jeromelaban) [inline src/Uno.UI/DataBinding/BindingPropertyHelper.FastConvert.cs:346]: split this into different methods, there's also already a FastStringToDouble converter below.

File: src/Uno.UI/DataBinding/BindingPropertyHelper.FastConvert.cs [hunk 2]
```diff
@@ -334,6 +339,240 @@ private static bool FastStringConvert(Type outputType, string input, ref object
...
+							return true;
+						case '9':
+							output = 9d;
+							return true;
+					}
+				}
+
+				var trimmed = input.Trim();
+
+				if (trimmed.Equals("nan", StringComparison.InvariantCultureIgnoreCase)
```
Comment (Xiaoy312) [inline src/Uno.UI/DataBinding/BindingPropertyHelper.FastConvert.cs:395]: Shouldn't these be case-sensitive?
> In addition to the numeric values, text syntax for x:Double permits the tokens `Infinity`, `-Infinity`, and `NaN`. These tokens are treated as case sensitive.
> https://docs.microsoft.com/en-us/dotnet/desktop/xaml-services/types-for-primitives#xdouble

It seems like the only exception is "Auto": (for wpf, but it seems to be still valid for uwp)
> In XAML you set the value to the string "Auto" (case insensitive) to enable the auto sizing behavior.
> https://docs.microsoft.com/en-us/dotnet/api/system.windows.frameworkelement.width?redirectedfrom=MSDN&view=net-5.0#remarks

Comment (carldebilly) [reply src/Uno.UI/DataBinding/BindingPropertyHelper.FastConvert.cs:395]: That's the result of the tests I did on UWP, not the documentation :-)

File: src/Uno.UI/DataBinding/BindingPropertyHelper.FastConvert.cs [hunk 3]
```diff
@@ -628,12 +633,58 @@ private static bool FastStringToGridLengthConvert(Type outputType, string input,
...
+
+				if (input.Length == 1)
+				{
+					// Fast path for one digit string-to-double.
+					// Often use in VisualStateManager to set double values (like opacity) to zero or one...
+					var c = input[0];
+					if (c >= '0' && c <= '9')
+					{
+						output= (double)(c - '0');
+							return true;
```
Comment (dr1rrb) [inline src/Uno.UI/DataBinding/BindingPropertyHelper.FastConvert.cs:654]: Alignment issue

File: src/Uno.UI/DataBinding/BindingPropertyHelper.FastConvert.cs [hunk 4]
```diff
@@ -628,12 +633,58 @@ private static bool FastStringToGridLengthConvert(Type outputType, string input,
...
+				}
+
+				if (input.Length == 1)
+				{
+					// Fast path for one digit string-to-double.
+					// Often use in VisualStateManager to set double values (like opacity) to zero or one...
+					var c = input[0];
+					if (c >= '0' && c <= '9')
+					{
+						output= (double)(c - '0');
```
Comment (dr1rrb) [inline src/Uno.UI/DataBinding/BindingPropertyHelper.FastConvert.cs:653]: needs space

File: src/Uno.UI/DataBinding/BindingPropertyHelper.FastConvert.cs [hunk 5]
```diff
@@ -628,12 +633,58 @@ private static bool FastStringToGridLengthConvert(Type outputType, string input,
...
+							return true;
+					}
+				}
+
+				var trimmed = input.Trim();
+
+				if (trimmed.Equals("nan", StringComparison.InvariantCultureIgnoreCase)
```
Comment (dr1rrb) [inline src/Uno.UI/DataBinding/BindingPropertyHelper.FastConvert.cs:660]:
```suggestion
				if (trimmed.Equals("nan", StringComparison.OrdinalIgnoreCase)
```

File: src/Uno.UI/DataBinding/BindingPropertyHelper.FastConvert.cs [hunk 6]
```diff
@@ -628,12 +633,58 @@ private static bool FastStringToGridLengthConvert(Type outputType, string input,
...
+					|| trimmed.Equals("auto", StringComparison.OrdinalIgnoreCase))
+				{
+					output = double.NaN;
+					return true;
+				}
+
+				if (trimmed.Equals("-infinity", StringComparison.InvariantCultureIgnoreCase))
```
Comment (dr1rrb) [inline src/Uno.UI/DataBinding/BindingPropertyHelper.FastConvert.cs:667]:
```suggestion
				if (trimmed.Equals("-infinity", StringComparison.OrdinalIgnoreCase))
```

File: src/Uno.UI/DataBinding/BindingPropertyHelper.FastConvert.cs [hunk 7]
```diff
@@ -628,12 +633,58 @@ private static bool FastStringToGridLengthConvert(Type outputType, string input,
...
+				{
+					output = double.NegativeInfinity;
+					return true;
+				}
+
+				if (trimmed.Equals("infinity", StringComparison.InvariantCultureIgnoreCase))
```
Comment (dr1rrb) [inline src/Uno.UI/DataBinding/BindingPropertyHelper.FastConvert.cs:673]:
```suggestion
				if (trimmed.Equals("infinity", StringComparison.OrdinalIgnoreCase))
```

**Review by dr1rrb — APPROVED** [review]

**Review by jeromelaban — APPROVED** [review]

**Pull request merged**